File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/SpecificArtistIterator.cs
namespace IteratorDesignPattern.Example3;

public class SpecificArtistIterator : IIterator<Song>
{
    private readonly List<Song> _songs;
    private int _currentIndex = 0;

    public SpecificArtistIterator(ArtistLibrary library, string artist)
    {
        _songs = library.GetArtistSongs().ContainsKey(artist)
            ? library.GetArtistSongs()[artist]
            : new List<Song>();
    }

    public bool HasNext()
    {
        return _currentIndex < _songs.Count;
    }

    public Song Next()
    {
        return _songs[_currentIndex++];
    }

    public void Reset()
    {
        _currentIndex = 0;
    }
}

File: java/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example3/src/main/java/behavioral/iterator/example3/SpecificArtistIterator.java
package behavioral.iterator.example3;

import java.util.*;

public class SpecificArtistIterator implements IIterator<Song> {
    private final List<Song> songs;
    private int currentIndex = 0;

    public SpecificArtistIterator(ArtistLibrary library, String artist) {
        songs = library.getArtistSongs().getOrDefault(artist, Collections.emptyList());
    }

    @Override
    public boolean hasNext() { return currentIndex < songs.size(); }

    @Override
    public Song next() { return songs.get(currentIndex++); }

    @Override
    public void reset() { currentIndex = 0; }
}
